frontend/catalyst/pennylane_extensions.py

```diff
@@ -2250,3 +2252,211 @@ def _callee():
+
+
+def vmap(
+    fn: Callable,
+    in_axes: Union[int, Sequence[Any]] = 0,
+    out_axes: Union[int, Sequence[Any]] = 0,
+    axis_size: Optional[int] = None,
+) -> Callable:
+    """A :func:`~.qjit` compatible vectorizing map for PennyLane/Catalyst.
+
+    Args:
+        f (Callable): A Python function containing PennyLane quantum operations.
+        in_axes (Union[int, Sequence[Any]]): It specifies the value(s) over which input
+            array axes to map.
+        out_axes (int): It specifies where te mapped axis should appear in the output.
+        axis_size (int): An integer can be optionally provided to indicate the size of the
+            axis to be mapped. If omitted, the size of the mapped axis will be inferred from
+            the provided arguments.
+
+    Returns:
+        (Callable): Vectorized version of ``fn``.
+
+    Raises:
+        TypeError: Invalid `in_axes` and `out_axes` values.
+        ValueError: Vectorize with more than one argument.
+        RuntimeError: Incompatible length of ``in_axes`` and positional arguments of ``fn``.
+    """
+
+    # Dispatch to jax.vmap when it is called outside qjit.
+    if not EvaluationContext.is_tracing():
+        return jax.vmap(fn, in_axes, out_axes)
+
+    # Check the validity of arguments
+    if isinstance(in_axes, list):
+        in_axes = tuple(in_axes)
+
+    if not all(isinstance(l, int) for l in tree_leaves(in_axes)):
+        raise TypeError(
+            "Invalid 'in_axes'; it can be an int or a tuple of "
+            f"PyTrees with integer leaves, but got {in_axes}"
+        )
+
+    if not all(isinstance(l, int) for l in tree_leaves(out_axes)):
+        raise TypeError(
+            f"Invalid 'out_axes'' it can be a PyTree with integer leaves, but got {out_axes}"
+        )
+
+    def batched_fn(*args, **kwargs):
+        """Vectorization wrapper around the hybrid program using catalyst.for_loop"""
+
+        # Check the validity of the input arguments
+        if isinstance(in_axes, tuple) and len(in_axes) != len(args):
+            raise TypeError(
+                "Invalid 'in_axes'; it can be an int or a tuple of PyTrees "
+                f"corresponding to the arguments, but got {len(in_axes)} and {len(args)}"
+            )
+
+        args_flat, args_tree = tree_flatten(args)
+        in_axes_flat = _vmap_tree_flatten(in_axes)
+
+        # import pdb
+        # pdb.set_trace()
+
+        batch_size = _get_batch_size(args_flat, in_axes_flat, axis_size)
+        batch_loc = _get_batch_loc(in_axes_flat)
+
+        # If vmap is requested on any axes
+        if len(batch_loc):
+            for loc in batch_loc:
+                # Support vectorization on dim > 0
+                if ax := in_axes_flat[loc]:
+                    up_axes = [*range(args_flat[loc].ndim)]
+                    up_axes[ax], up_axes[0] = up_axes[0], up_axes[ax]
+                    args_flat[loc] = jnp.transpose(args_flat[loc], up_axes)
```

Comment:
What is the purpose of this block?

Reply:
To enable vectorization over higher dimensions, we must remap the dimensions of the provided batch arguments by transposing them before iterating over them in the for-loop

Imagine the following example,

``` python
@qjit
def workflow(y, x):
    @qml.qnode(qml.device(backend, wires=1))
    def circuit(y, x):
        qml.RX(jnp.pi * x[0] * y, wires=0)
        qml.RY(x[1] ** 2, wires=0)
        qml.RX(x[1] * x[2] * y, wires=0)
        return qml.expval(qml.PauliZ(0))

    res1 = vmap(circuit, in_axes=[None, 1])(y[0], x)
    res2 = vmap(circuit, in_axes=(0, 1))(y, x)
    return res1 , res2

x = jnp.array(
    [
        [0.1, 0.4],
        [0.2, 0.5],
        [0.3, 0.6],
    ]
)

y = jnp.array([1, 2])

result = workflow(y, x)
```

The user provides `x` and they can specify the axis over which vmap applies the mapping. Here, in this example, they provide `in_axes = (*, 1)` and it really means that the for_loop will iterate over 1-axis of `x`. We cannot directly use this list of arguments in our pipeline, and to remap it, I used `jnp.transpose` to "efficiently?!" remap the batched arguments and call for_loop over the updated batch arguments.

Reply:
The issue is not related to the numpy style indexing but the implementation of this indexing in catalyst.vmap. I found a better solution using `jax.numpy.take` to make the implementation similar to the numpy style indexing. Thanks!